utils/s2n_result.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#pragma once
+
+#include <s2n.h>
+#include <stdbool.h>
+
+/* A value which indicates the outcome of a function */
+typedef struct {
+    int __error_signal;
+} s2n_result;
+
+/* used to signal a successful function return */
+extern const s2n_result S2N_RESULT_OK;
+/* used to signal an error while executing a function */
+extern const s2n_result S2N_RESULT_ERROR;
+
+#if defined(__clang__) || defined(__GNUC__)
+#define S2N_RESULT_MUST_USE __attribute__((warn_unused_result))
+#else
+#define S2N_RESULT_MUST_USE
+#endif
+
+/* returns true when the result is S2N_RESULT_OK */
+S2N_RESULT_MUST_USE bool s2n_result_is_ok(s2n_result result);
+
+/* returns true when the result is S2N_RESULT_ERROR */
+S2N_RESULT_MUST_USE bool s2n_result_is_error(s2n_result result);
+
+/* used in function declarations to signal function fallibility */
+#define S2N_RESULT S2N_RESULT_MUST_USE s2n_result
+
+/* s2n_result GUARD helpers */
+/* note: eventually this will just alias GUARD and be deprecated once everything is using s2n_result */
+#define GUARD_RESULT( x )               do {if ( s2n_result_is_error(x) ) return S2N_RESULT_ERROR;} while (0)
+#define GUARD_POSIX_AS_RESULT( x )      do {if ( (x) < 0 ) return S2N_RESULT_ERROR;} while (0)
+#define GUARD_RESULT_AS_POSIX( x )      do {if ( s2n_result_is_error(x) ) return S2N_FAILURE;} while (0)
```

Comment:
I generally prefer more explicit function names as you have it here, but if we're going to be calling these macro 100's of times throughout the codebase, I think reducing the verbosity of each Macro call may be worthwhile. Since `S2N_RESULT->POSIX` and `POSIX->S2N_RESULT` will be the two most common translation macros (until we're fully migrated), we should be able to drop the input type and leave only the output type in the Macro Name. 

What do others think about these names:
`GUARD_AS_RESULT()` // Implicitly takes POSIX result as input
`GUARD_AS_POSIX()` // Implicitly takes S2N_RESULT result as input